WIP: Try to implement accepting nullable params

coltfred opened the pull request:
Opening PR to see how it builds and get any feedback I can from @Dushistov.

It's unclear to me when SwigFrom on jobject should be used vs SwigInto. I'm following the model that was used for f64, et al.

Dushistov commented on macroslib/src/java_jni/jni-include.rs:
```diff
@@ -1361,3 +1361,46 @@ impl<T: SwigForeignClass> SwigFrom<Option<T>> for jobject {
     }
 }
+
+#[swig_from_foreigner_hint = "java.util.Optional<T>"]
```
It is not Java way to pass `java.util.Optional<T>` as input argument, see for example:

https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments
http://dolszewski.com/java/java-optional-parameters/

The problem is three states instead of two in Rust: analog of `Some`, analog of `None` and `null`.
For output parameters complexity that add null compensated by methods like `Optional.flatMap` and so on things.
But for input parameters it should be just `T`, because of any `T` can be `null`.

coltfred replied:
Fair enough. Made it T.

Dushistov commented on macroslib/src/java_jni/jni-include.rs:
```diff
@@ -1361,3 +1361,46 @@ impl<T: SwigForeignClass> SwigFrom<Option<T>> for jobject {
+        if x.is_null() {
+            None
+        } else {
+            let x = unsafe { (**env).NewLocalRef.unwrap()(env, x) };
```
the second call of `NewLocalRef` looks strange.

Dushistov commented on macroslib/src/java_jni/jni-include.rs:
```diff
@@ -1361,3 +1361,46 @@ impl<T: SwigForeignClass> SwigFrom<Option<T>> for jobject {
+                None
+            } else {
+
+                let class_id = <T>::jni_class_name();
```
In fact all things like this is expensive (from CPU usage point of view): FindClass, CallMethod etc.

Better would be partly handle it in https://github.com/Dushistov/rust_swig/blob/master/macroslib/src/java_jni/map_type.rs with help of Java not JNI.

coltfred replied:
I don't see anywhere else where that happens. Are there other examples of what you're talking about? I made it work like the other example of `Vec<T>` for now. While inefficient, it seems to work for null and non null cases.

coltfred commented:
@Dushistov I've added tests and fixed it up to not take optional. Still has the expensive code in it because I'm not sure how to avoid that, but :man_shrugging: .

Dushistov commented:
> Still has the expensive code in it because I'm not sure how to avoid that, but

I took your commits and mine, see #165. This demonstrate how it is possible to work with 
fields without `SetField` and so on methods.

The DSL for type conversation is still not mature enough to handle all cases, so I have to add Rust code to `macroslib/src/java_jni/map_type.rs`,
this should help with adding proper `@Nullable` annotation to input parameters.
